tests/src/test/java/org/eclipse/hono/tests/client/ClientTestHelper.java

```diff
@@ -0,0 +1,115 @@
+/**
+ * Copyright (c) 2018 Bosch Software Innovations GmbH and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ */
+package org.eclipse.hono.tests.client;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+import org.apache.qpid.proton.message.Message;
+import org.eclipse.hono.client.MessageConsumer;
+import org.eclipse.hono.client.MessageSender;
+import org.eclipse.hono.tests.IntegrationTestSupport;
+import org.eclipse.hono.util.MessageHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.ext.unit.TestContext;
+
+/**
+ * Base class for integration tests for Hono's AMQP 1.0 services.
+ */
+public abstract class ClientTestHelper {
+
+    /**
+     * A logger to be used by subclasses.
+     */
+    protected final Logger LOGGER = LoggerFactory.getLogger(getClass());
+
+    private static final long   DEFAULT_TEST_TIMEOUT = 15000; // ms
+
+    /**
+     * Creates a test specific message consumer.
+     *
+     * @param tenantId        The tenant to create the consumer for.
+     * @param messageConsumer The handler to invoke for every message received.
+     * @return A future succeeding with the created consumer.
+     */
+    protected abstract Future<MessageConsumer> createConsumer(String tenantId, Consumer<Message> messageConsumer);
```

Comment:
this method is never invoked from this class, is it?

Reply:
Yes true! i figured by putting in the parent class, we avoid duplicating code. isn't that better than defining the methods in both subclasses?

Reply:
Well, FMPOV it only makes sense to declare abstract methods if those methods are use/required by some code in the same class that declares them. In your case it would make sense to declare the methods in the abstract class, if all test cases were defined in the abstract class as well and would use those methods to create the (subclass specific) senders and consumers. The subclasses would then e.g. be used to provide variants of sender and consumer and maybe implement additional test cases.

On the other hand, why should a subclass be forced to implement methods that are not used anywhere? So, unless you redesign the abstract class to actually invoke those methods, they should not be declared there IMHO.

Reply:
i see your point. will update the PR... thanks!